src/sentry/api/endpoints/organization_stats_v2.py

```diff
@@ -0,0 +1,225 @@
+    def get(self, request, project):
+        start, end, rollup = get_date_range_rollup_from_params(request.GET, "1h", round_range=True)
+
+        # TODO: see if there's a better way to get the user's projects
+        project_list = []
+        team_list = Team.objects.get_for_user(organization=project.organization, user=request.user)
+        for team in team_list:
+            project_list.extend(Project.objects.get_for_user(team=team, user=request.user))
+        project_ids = list({p.id for p in project_list})
+
+        if project.id not in project_ids:
+            return Response(status=404)
+
+        result = outcomes.query(
+            start=start,
+            end=end,
+            rollup=rollup,
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [project.organization.id], "project_id": [project.id]},
+            orderby=["time"],
+        )
+
+        response = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+        }
+        for row in result:
+            nested_update(
+                response[CATEGORY_NAME_MAP[row["category"]]][row["time"]], map_row_to_format(row)
+            )
+
+        response = {
+            category: zerofill(list(val.values()), start, end, rollup, "time")
+            for category, val in response.items()
+        }
+        return Response(response)
+
+
+def outcome_to_string(outcome):
+    # TODO: why do we rename this?
+    return "dropped" if outcome == Outcome.RATE_LIMITED else Outcome(outcome).api_name()
+
+
+def rate_limited_reason_mapper(reason):
+    # billing logic ported over. TODO: combine usages of this into some abstracted module?
+    if reason in {"usage_exceeded", "grace_period"}:
+        reason_val = "overQuota"
+    elif reason == "smart_rate_limit":
+        reason_val = "spikeProtection"
+    else:
+        reason_val = "other"
+    return reason_val
+
+
+# TODO: add ts and date fields to replace "timestamp" field
+
+
+def map_row_to_format(row):
+    obj = {"time": row["time"]}
+    if row["outcome"] == Outcome.RATE_LIMITED:
+        obj[outcome_to_string(row["outcome"])] = {
+            rate_limited_reason_mapper(row["reason"]): {
+                "quantity": row["quantity"],
+                "times_seen": row["times_seen"],
+            }
+        }
+    else:
+        obj[outcome_to_string(row["outcome"])] = {
+            "quantity": row["quantity"],
+            "times_seen": row["times_seen"],
+        }
+    return obj
+
+
+def nested_update(d, u):
+    # https://stackoverflow.com/a/3233356
+    for k, v in u.items():
+        if isinstance(v, collections.abc.Mapping):
+            d[k] = nested_update(d.get(k, {}), v)
```

Comment:
Recursing to arbitrary depth seems squirrelly to me. In the places where this is called, would it be possible to unpack the appropriate number of levels instead?

It looks like this is meant to explore a structure like
```python
{
    "accepted": {"quantity": 0, "times_seen": 0},
    "filtered": {"quantity": 0, "times_seen": 0},
    "dropped": {
        "overQuota": {"quantity": 0, "times_seen": 0},
        "spikeProtection": {"quantity": 0, "times_seen": 0},
        "other": {"quantity": 0, "times_seen": 0},
    },
}
```
and hit a `{"quantity": 0, "times_seen": 0}` object whether it's one or two layers deep. See my other comment about introducing classes regarding that.